### PR TITLE
feat(data-management): Allow deleting event/property definitions

### LINKED ISSUE #7315: Allow cleaning up events and properties
## Is your feature request related to a problem?

As a team accumulates data, they will have events/properties that are no longer in use that clutter their app.

Example from our team:
image

## Describe the solution you'd like

Under event/properties stats allow 'hiding' events/properties from view.

## Describe alternatives you've considered

Alternatively, go the route of "verified" events/properties (as He@p) does - show those first. See also https://help.heap.io/category/heap-administration/data-governance/

## Additional context

Note: Existing reports, etc should still work well even with 'removed' events/properties.

Related: https://github.com/PostHog/posthog/issues/7126

### PR DESCRIPTION
**Pull request by Twixes**
## Problem

It wasn't possible to remove junk event or property definitions from PostHog, which was annoying – speaking both from personal experience and user feedback, [such as this](https://posthogusers.slack.com/archives/C01GLBKHKQT/p1679518041534929):

> Is it possible to hide, archive or otherwise delete those unwanted properties?
> Seems like something you should be able to do - if a property accidentally (or even maliciously) gets sent to a project, will it just sit there forever, polluting it?

## Changes

Resolves #7315 by making definitions deletable:

button
event
property

## How did you test this code?

Added a couple API tests.

### CODE REVIEW

File: frontend/src/scenes/data-management/definition/definitionLogic.ts [hunk 1]
```diff
@@ -85,6 +87,20 @@ export const definitionLogic = kea<definitionLogicType>([
 
                     return definition
                 },
+                deleteDefinition: async () => {
+                    if (values.isEvent) {
+                        await api.eventDefinitions.delete({ eventDefinitionId: values.definition.id })
+                    } else {
+                        await api.propertyDefinitions.delete({ propertyDefinitionId: values.definition.id })
```
Comment (tiina303) [inline frontend/src/scenes/data-management/definition/definitionLogic.ts:94]: this works for person properties too?

File: frontend/src/scenes/data-management/definition/definitionLogic.ts [hunk 2]
```diff
@@ -85,6 +87,20 @@ export const definitionLogic = kea<definitionLogicType>([
...
+                    }
+                    router.actions.push(values.isEvent ? urls.eventDefinitions() : urls.propertyDefinitions())
+                    if (values.isEvent) {
+                        eventDefinitionsTableLogic.findMounted()?.actions.loadEventDefinitions()
+                    } else {
+                        propertyDefinitionsTableLogic.findMounted()?.actions.loadPropertyDefinitions()
+                    }
```
Comment (tiina303) [inline frontend/src/scenes/data-management/definition/definitionLogic.ts:101]: don't we in other places where we do something similar just update the frontend local data (e.g. person property updates update the local data by adding / removing the property, which is more efficient than reloading everything - nicer experience if you're deleting many things and reloading is slow)?

**Review by tiina303 — COMMENTED** [review]
Lets make sure we log this action in the activity log (e.g. if we need to rebuild the properties table we should also replay the deletes)

**Review by pauldambra — APPROVED** [review]
Does what it says on the tin... 👍 

Should someone be able to delete a PostHog event like autocapture or $lib? Should that be restricted to staff?

Should someone be able to delete a verified event without un-verifying first... or 🤷 

I like the activity log point. Data management is already a tabbed UI so there's somewhere to put the logs. We don't have this type of activity log but it is relatively straight-forward to add them...

File: frontend/src/scenes/data-management/definition/definitionLogic.ts (hunk 1 above)
Comment (Twixes) [reply frontend/src/scenes/data-management/definition/definitionLogic.ts:94]: Yup. Well, property definitions are currently shared among events and persons.

File: frontend/src/scenes/data-management/definition/definitionLogic.ts (hunk 2 above)
Comment (Twixes) [reply frontend/src/scenes/data-management/definition/definitionLogic.ts:101]: We do, I actually wanted to do that first here… But the event/property definition table logics have some custom-ish pagination handling (e.g. there's a next/previous page URL normalization process), which I don't fully understand, so just reloading completely to make sure there's no stale data – which is worse than having to make an extra request.

**Comment by Twixes** [comment]
Good questions @pauldambra. I considered whether deletion of "official" definitions should be possible, and IMO yes. E.g. if you're only tracking a mobile app, you might not want to see "Pageview" at all – why would you? And if you ever do start tracking pageviews, the definition will simply be recreated.
I didn't actually consider verified events, but I don't think that makes a significant difference anyway. Same as above, it's just the definition, you're not losing actual data by deleting it.

As for the activity log, it's a good point, but I'm gonna bias towards just shipping this, given that we don't already have an activity log for definitions. It'd be nice, but this alone seems to alleviate more burning user pain.

File: frontend/src/scenes/data-management/definition/definitionLogic.ts [hunk 3]
```diff
@@ -93,6 +94,10 @@ export const definitionLogic = kea<definitionLogicType>([
                     } else {
                         await api.propertyDefinitions.delete({ propertyDefinitionId: values.definition.id })
                     }
+                    posthog.capture(`${values.singular} definition deleted`, {
```
Comment (tiina303) [inline frontend/src/scenes/data-management/definition/definitionLogic.ts:97]: this should be in the backend - folks can hit the api above directly too

Comment (Twixes) [reply frontend/src/scenes/data-management/definition/definitionLogic.ts]: Fair, moved

File: posthog/api/property_definition.py [hunk 4]
```diff
@@ -474,3 +476,10 @@ def get_object(self):
     @extend_schema(parameters=[PropertyDefinitionQuerySerializer])
     def list(self, request, *args, **kwargs):
         return super().list(request, *args, **kwargs)
+
+    def destroy(self, request: request.Request, *args: Any, **kwargs: Any) -> response.Response:
+        instance = self.get_object()
+        self.perform_destroy(instance)
+        # Casting, since an anonymous use CANNOT access this endpoint
+        report_user_action(cast(User, request.user), "property definition deleted", {"name": instance.name})
```
Comment (tiina303) [inline posthog/api/property_definition.py:484]: could you add type in addition to "name" as well for easier usage later, it's helpful to know which 3 it is from https://github.com/PostHog/posthog/blob/016c49cb41d9dd708b55de287bb466a4996303c2/posthog/models/property_definition.py#L31-33

**Review by tiina303 — APPROVED** [review]

File: posthog/api/property_definition.py (hunk 4 above)
Comment (Twixes) [reply posthog/api/property_definition.py]: Done

**Pull request merged**